torch/_dynamo/variables/misc.py

```diff
@@ -363,8 +363,10 @@ def visit(node):
                 self.fn_cls.setup_context
                 != torch.autograd.function._SingleLevelFunction.setup_context
             ):
-                unimplemented(
-                    "NYI - autograd.Function with custom setup_context method"
+                self.fn_cls.forward = (
+                    torch.autograd.function.autograd_function_forward_rewritten(
+                        self.fn_cls.forward, self.fn_cls.setup_context
+                    )
```

Comment:
I don't think this is the right way, since we modified user's code. However, I still don't figure out a proper way to handle this. The main challenge is the new generated forward function has ```__closure__``` but the original forward doesn't have. This would cause guard failures as forward function should not have guards on ```__closure__```. I was trying to use sth like:
```
new_forward = types.FunctionType(
        new_forward.__code__,
        original_forward.__globals__,
        original_forward.__name__,
        original_forward.__defaults__,
        original_forward.__closure__,
    )
```
to update this function, but it doesn't work. The length of the new forward function ```__clousure__``` is 2, so we are not allowed to update it as None.

cc @zou3519